File: 202.cpp
// https://leetcode.com/problems/happy-number/
class Solution {
 public:
  bool isHappy(int n) {
    std::unordered_set<int> visited;
    while (n != 1 && !visited.count(n)) {
      visited.insert(n);
      int tmp = 0;
      while (n) {
        tmp += (n % 10) * (n % 10);
        n /= 10;
      }
      n = tmp;
    }
    return n == 1;
  }
};

File: golang/202.go
package main

func isHappy(n int) bool {
	visited := make(map[int]int)
	for n != 1 && visited[n] == 0 {
		visited[n] += 1
		tmp := 0
		for n != 0 {
			tmp += (n % 10) * (n % 10)
			n /= 10
		}
		n = tmp
	}
	return n == 1
}
